File: trie/trie.go
// Package trie is word dictionary implementation using trie in go
package trie

// trieNode saves trie structure
type trieNode struct {
	childMap map[rune]*trieNode
	isEnd    bool
}

// Trie struct contains data and methods
type Trie struct {
	root      *trieNode
	wordCount int
}

// NewTrie creates new instance of trie
func NewTrie() *Trie {
	return &Trie{
		root: &trieNode{
			childMap: map[rune]*trieNode{},
			isEnd:    false,
		},
		wordCount: 0,
	}
}

// Size returns word count in trie
func (pr *Trie) Size() int {
	return pr.wordCount
}

// Insert inserts new word in trie
// returns false if this word already present
func (pr *Trie) Insert(word string) bool {
	node := pr.root
	for _, ch := range word {
		if newNode, ok := node.childMap[ch]; ok {
			node = newNode
		} else {
			node.childMap[ch] = &trieNode{
				childMap: map[rune]*trieNode{},
				isEnd:    false,
			}
			node = node.childMap[ch]
		}
	}
	if node.isEnd {
		return false
	}
	node.isEnd = true
	pr.wordCount++
	return true
}

// InsertAll inserts all words in trie returns number of words actually
// inserted.
func (pr *Trie) InsertAll(words []string) int {
	var res = 0
	for _, word := range words {
		if pr.Insert(word) {
			res++
		}
	}
	return res
}

// Retrieve retrieves all words in trie starting with prefix.
func (pr *Trie) Retrieve(prefix string) []string {
	node, depth := longestMatch(prefix, pr.root)
	if depth != len(prefix) {
		return []string{}
	}
	return allChild(prefix, node)
}

// Contains checks if given word is in trie.
func (pr *Trie) Contains(word string) bool {
	node, depth := longestMatch(word, pr.root)
	return depth == len(word) && node.isEnd
}

// Delete deletes given word from trie. Returns false if word is not in trie.
func (pr *Trie) Delete(word string) bool {
	node, depth := longestMatch(word, pr.root)
	if depth == len(word) && node.isEnd {
		node.isEnd = false
		return true
	}
	return false
}

func longestMatch(prefix string, root *trieNode) (*trieNode, int) {
	node := root
	var depth = 0
	for _, ch := range prefix {
		if newNode, ok := node.childMap[ch]; ok {
			node = newNode
			depth++
		} else {
			return node, depth
		}
	}
	return node, depth
}

func allChild(prefix string, node *trieNode) []string {
	var res []string
	if node.isEnd {
		res = append(res, prefix)
	}

	for ch, childNode := range node.childMap {
		newStr := prefix + string(ch)
		res = append(res, allChild(newStr, childNode)...)
	}
	return res
}

File: trie/trie.py
""" Trie is an efficient information reTrieval data structure. Using Trie,
    search complexities can be brought to optimal limit (key length). If we
    store keys in binary search tree, a well balanced BST will need time
    proportional to M * log N, where M is maximum string length and N is
    number of keys in tree. Using Trie, we can search the key in O(M) time.
    However the penalty is on Trie storage requirements
"""

class Node:
    """ Node contains character and it's children trie node """
    def __init__(self):
        self.children = [None] * 26
        self.word_finished = False


class Trie:
    """ Trie Data Structure implementation """
    def __init__(self):
        self.root = Node()

    def char_to_index(self, char):
        """ converts key character into index
            use only 'a' through 'z' lower case
        """
        return ord(char) - ord('a')

    def insert(self, word):
        """ inserts a word in the trie data structure"""
        current = self.root
        for character in word:
            index = self.char_to_index(character)
            if not current.children[index]:
                current.children[index] = Node()
            current = current.children[index]
        current.word_finished = True

    def search(self, word):
        """ returns true if the complete word is present in trie,
            else false
        """
        current = self.root
        for character in word:
            index = self.char_to_index(character)
            if not current.children[index]:
                return False
            current = current.children[index]
        return current is not None and current.word_finished

    def starts_with(self, prefix):
        """ returns true if any word present in the trie start with the
            given prefix, else false
        """
        current = self.root
        for character in prefix:
            index = self.char_to_index(character)
            if not current.children[index]:
                return False
            current = current.children[index]
        return True

    def delete(self, word):
        """ deletes word by negating word_finished boolean variable """
        current = self.root
        for character in word:
            index = self.char_to_index(character)
            if not current.children[index]:
                raise Exception("word not present in trie")
            current = current.children[index]

        if current.word_finished:
            current.word_finished = False


def main():
    """ operational function """
    tree = Trie()
    tree.insert("apple")
    print("is apple present?", tree.search("apple")) # True
    print("is app present?", tree.search("app")) # False
    print("starts with app?", tree.starts_with("app")) # True

    print("inserting 'app' in trie")
    tree.insert("app")
    print("is app present?", tree.search("app")) # True

    print("removing 'app' from trie")
    tree.delete("app")
    print("is app present?", tree.search("app")) # False


if __name__ == "__main__":
    main()

